brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+        So we pre-calculate them. Because of the fact that these are reused,
+        it is in principle possible to update the fitting as new data come in,
+        by just incrementally adding the products of new data and
+        their corresponding part of design matrix
+        """
+        if scan_onsets is None:
+            # assume that all data are acquired within the same scan.
+            D = np.diag(np.ones(n_T - 1), -1) + np.diag(np.ones(n_T - 1), 1)
+            F = np.eye(n_T)
+            F[0, 0] = 0
+            F[n_T - 1, n_T - 1] = 0
+        else:
+            # Each value in the scan_onsets tells the index at which
+            # a new scan starts. For example, if n_T = 500, and
+            # scan_onsets = [0,100,200,400], this means that the time points
+            # of 0-99 are from the first scan, 100-199 are from the second,
+            # 200-399 are from the third and 400-499 are from the fourth
+            run_TRs = np.diff(np.append(scan_onsets, n_T))
+            logger.debug('I infer that the number of volumes'
+                         ' in each scan are: {}'.format(run_TRs))
+
+            D_ele = list(map(self._D_gen, run_TRs))
+            F_ele = list(map(self._F_gen, run_TRs))
+            D = []
+            for d_ele in D_ele:
+                D = scipy.linalg.block_diag(D, d_ele)
+            F = []
+            for f_ele in F_ele:
+                F = scipy.linalg.block_diag(F, f_ele)
+            # D and F above are templates for constructing
+            # the inverse of temporal covariance matrix of noise
+
+        XTY = np.dot(design.T, Y)
+        XTDY = np.dot(np.dot(design.T, D), Y)
+        XTFY = np.dot(np.dot(design.T, F), Y)
+
+        YTY_diag = np.zeros([np.size(Y, axis=1)])
+        YTDY_diag = np.zeros([np.size(Y, axis=1)])
+        YTFY_diag = np.zeros([np.size(Y, axis=1)])
+        for i_V in range(n_V):
+            YTY_diag[i_V] = np.dot(Y[:, i_V].T, Y[:, i_V])
+            YTDY_diag[i_V] = np.dot(np.dot(Y[:, i_V].T, D), Y[:, i_V])
+            YTFY_diag[i_V] = np.dot(np.dot(Y[:, i_V].T, F), Y[:, i_V])
+
+        XTX = np.dot(design.T, design)
+        XTDX = np.dot(np.dot(design.T, D), design)
+        XTFX = np.dot(np.dot(design.T, F), design)
+        return XTY, XTDY, XTFY, YTY_diag, YTDY_diag, YTFY_diag, XTX, XTDX, XTFX
+
+    def calc_dist2_GP(self, coords=None, inten=None,
+                      GP_space=False, GP_inten=False):
+        # calculate the square of difference between each voxel's location
+        # coorinates and image intensity.
+        if GP_space:
+            assert coords is not None, 'coordinate is not provided'
+            n_V = coords.shape[1]
+            dist2 = np.zeros([n_V, n_V])
+            # square of spatial distance between every two voxels
+            for dim in range(np.size(coords, 1)):
+                coord_tile = np.tile(coords[:, dim], [n_V, 1])
+                dist2 = dist2 + (coord_tile - coord_tile.T)**2
```

Comment:
Also: just ran into a similar problem elsewhere and I wonder if http://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html will do what you want without reinventing the wheel?